exes/ethercat/inc/public/tfc/ec/devices/schneider/atv320.hpp

```diff
@@ -396,9 +397,67 @@ using acceleration_ramp_time_ACC = setting<ecx::index_t{ 0x203c, 0x02 }, "ACC",
 // 100 = 10 seconds
 // 10 = 1 second
 using deceleration_ramp_time_DEC = setting<ecx::index_t{ 0x203c, 0x03 }, "DEC", "Deceleration time ramp", deciseconds, 1>;
-}  // namespace tfc::ec::devices::schneider
 
-namespace tfc::ec::devices::schneider {
+namespace detail {
+struct speed {
+  decifrequency value{ 0 * dHz };
+  bool reverse{ false };
+  constexpr auto operator==(speed const& other) const noexcept -> bool = default;
+};
+constexpr auto percentage_to_deci_freq(mp_units::quantity<mp_units::percent, double> percentage,
+                                       [[maybe_unused]] low_speed_LSP min_freq,
+                                       [[maybe_unused]] high_speed_HSP max_freq) noexcept -> speed {
+  if (percentage == 0 * mp_units::percent) {
+    return { .value = 0 * dHz, .reverse = false };
+  }
+  bool reverse = percentage.numerical_value_ < 0;
+  auto mapped{ ec::util::map(mp_units::abs(percentage).numerical_value_, (1.0 * mp_units::percent).numerical_value_,
+                             (100.0 * mp_units::percent).numerical_value_, min_freq.value.numerical_value_,
+                             max_freq.value.numerical_value_) };
```

Comment:
Do you need to throw away the types?

Reply:
since we are mapping cross dimensions, yes. The correct way would be to ask mp to provide map function in his math file

Reply:
Or write explicit mapping function for mp

Reply:
I think that our static_cast to double might be messing it up but the from type should cancel out with the division in map.

Reply:
what was the reason for static_cast to double, do you remember?

Reply:
mp-units uses the underlying types for calculations. You can get values far from what you expect. We can experiment with changing it back though or maybe cast to double and then cast back.

Reply:
I will tinker with removing the casting and see if it provides same results test wise

Reply:
now explicitly for double (implicit casting), and mp units truncating to double